Group network configurations by chain

mcmire opened the pull request:
## Explanation

<!--
Thanks for your contribution! Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?
* Are there any changes whose purpose might not obvious to those unfamiliar with the domain?
* If your primary goal was to update one package but you found you had to update another one along the way, why did you do so?
* If you had to upgrade a dependency, why did you do so?
-->

Currently, in the client, it is possible to have multiple networks with different RPC endpoint URLs representing the same chain. This creates a problem because if all we have is a chain ID, we don't know which URL to use for requests.

To solve this, we plan on consolidating the UX on the client side such that each network corresponds to exactly one chain. Users can then select which default RPC URL they'd like to use for requests. This commit implements the controller changes necessary to support this UX.

Here are some more details on the changes here:

- The concept of a network configuration has been repurposed such that instead of representing an RPC endpoint, it now represents a whole chain.

  - A network configuration may have multiple RPC endpoints, and one of them must be designated as the default.
  - Some RPC endpoints are special in that they represent Infura API URLs; these have the same object shape as "non-Infura" (custom) RPC endpoints, but the Infura project ID is hidden and injected into the RPC URL when creating the network client.
  - There is no longer a 1-to-1 relationship between network configuration and network client; rather, the 1-to-1 relationship exists between RPC endpoint and network client. This means that the ID of the network client which is created for an RPC endpoint is stored on that RPC endpoint instead of the whole network configuration.

- The `networkConfigurations` state property has been replaced with `networkConfigurationsByChainId`. This continues to be an object, but the data inside is organized such that network configurations are identified by chain ID instead of network client ID as they were previously.

- The methods `upsertNetworkConfiguration` and `removeNetworkConfiguration` have been removed. These methods always did more than simply add or remove a network configuration; they also updated the registry of network clients. Instead, these methods have been replaced with `addNetwork`, `updateNetwork`, and `removeNetwork`.

  - `addNetwork` creates new network clients for each RPC endpoint in the given network configuration.
  - `updateNetwork` takes a chain ID referring to a network configuration and a draft version of that network configuration, and adds or removes network clients for added or removed RPC
    endpoints.
  - `removeNetwork` takes a chain ID referring to a network configuration and removes the network clients for each of its RPC endpoints.

- In addition, due to the changes to network configuration itself, there are new restrictions on `networkConfigurationsByChainId`, which are validated on initialization and on update. These are:

  - The network controller cannot be initialized with an empty collection of network configurations. This is because there must be a selected network client so that consumers have a provider to use out of the gate.
  - Consequently, the last network configuration cannot be removed.
  - The network configuration that contains a reference to the currently selected network client cannot be removed.
  - The chain ID of a network configuration must match the same chain that it's filed under in `networkConfigurationsByChainId`.
  - No two network configurations can have the same chain ID.
  - A RPC endpoint in a network configuration must have a well-formed URL.
  - A network configuration cannot have duplicate RPC endpoints.
  - No two RPC endpoints (regardless of network configuration) can have the same URL. Equality is currently determined by normalizing URLs as per RFC 3986 and may include data like request headers in the future.
  - If a network configuration has an Infura RPC endpoint, its chain ID must match the set chain ID of the network configuration.
    - Changing the chain ID of a network configuration is possible, but any existing Infura RPC endpoint must be replaced with the one that matches the new chain ID.
  - No two RPC endpoints (regardless of network configuration) can have the same network client ID.

- Finally, the `trackMetaMetricsEvent` option has been removed from the constructor. This was previously used in `upsertNetworkConfiguration` to create a MetaMetrics event when a new network added, but I've added a new event `NetworkController:networkAdded` to allow the client to do this on its own accord.


## References

<!--
Are there any issues that this pull request is tied to? Are there other links that reviewers should consult to understand these changes better?

For example:

* Fixes #12345
* Related to #67890
-->

Fixes #4189.
Fixes #3793.

## Changelog

<!--
If you're making any consumer-facing changes, list those changes here as if you were updating a changelog, using the template below as a guide.

(CATEGORY is one of BREAKING, ADDED, CHANGED, DEPRECATED, REMOVED, or FIXED. For security-related issues, follow the Security Advisory process.)

Please take care to name the exact pieces of the API you've added or changed (e.g. types, interfaces, functions, or methods).

If there are any breaking changes, make sure to offer a solution for consumers to follow once they upgrade to the changes.

Finally, if you're only making changes to development scripts or tests, you may replace the template below with "None".
-->

(Updated in the PR.)

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

mcmire commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -63,32 +73,178 @@ export type NetworkMetadata = {
+  /**
+   * The name of the token that represents the native currency for the chain.
+   */
+  nativeTokenName: string;
```
This property used to be called `ticker`. I decided to rename it because although "ticker" makes sense from the perspective of trading, I feel that it's more of a colloquial term. I believe it's more technically accurate to say that each chain has a native token, and that token has a name. So that's why I decided to rename this. That said, if we feel that `ticker` isn't confusing, I'm okay with sticking with it. Thoughts?

MajorLift replied:
see https://github.com/MetaMask/core/pull/4286#discussion_r1630102068

Gudahtt replied:
It's not correct to call this a "token" though. It's a currency. The term we use for it elsewhere is "native currency".

We use the term "tokens" typically for on-chain assets like ERC-20, ERC-721, and ERC-1155. These all provide a "token" that is represented on-chain. But no such thing exists for currency. Similarly, we wouldn't describe real-life currency like USD as "tokens".

mcmire replied:
Change made here: [98d4c2a](https://github.com/MetaMask/core/pull/4286/commits/98d4c2a0269e9897453616c51610a66c3f308472)

mcmire replied:
Resolving this conversation, but feel free to reopen if you feel differently.

mcmire commented on packages/network-controller/CHANGELOG.md:
```diff
@@ -7,6 +7,41 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  - The `defaultRpcEndpointUrl` of a network configuration in `networkConfigurationsByChainId` must match an entry in its `rpcEndpoints`.
+  - `selectedNetworkClientId` must match the `networkClientId` of an RPC endpoint in `networkConfigurationsByChainId`.
+- **BREAKING:** Update `getNetworkConfigurationByNetworkClientId` so that when given an Infura network name (that is, a value from `InfuraNetworkType`), it will return a masked version of the RPC endpoint URL for the associated Infura network ([#4268](https://github.com/MetaMask/core/pull/4286))
+  - If you want the unmasked version, you'll need the `url` property from the network _client_ configuration, which you can get by calling `getNetworkClientById` and then accessing the `configuration` property off of the network client.
```
As an alternative to requiring this step, I could make `getNetworkConfigurationByNetworkClientId` (or any other methods that return a single network configuration or multiple network configurations) unmask the Infura API URL automatically. Thoughts?

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -63,32 +73,178 @@ export type NetworkMetadata = {
+   */
+  name: string;
+  /**
+   * The name of the token that represents the native currency for the chain.
```
RE https://github.com/MetaMask/core/pull/4286#discussion_r1612090728: using the comments to provide context from a non-technical POV seems like a good middle ground.
```suggestion
   * The name of the token that represents the native currency for the chain (i.e. the ticker symbol). 
```

mcmire replied:
Sure!

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -186,26 +295,35 @@ export type CustomNetworkClientId = string;
-export type NetworksMetadata = {
-  [networkClientId: NetworkClientId]: NetworkMetadata;
-};
+export type NetworksMetadata = Record<NetworkClientId, NetworkMetadata>;
 
```
nit: This is probably outside of the PR's scope, but `NetworkProperties` might be a possible replacement for `NetworkMetadata`, which could be confused with `NetworkStateMetadata`.

In general, we might want to look into categorizing our data types and properties so that they consistently distinguish between the following categories (and possibly more). `TransactionMeta`, for example, might benefit from being broken up into a few types:
- Static descriptors/metadata (e.g. `NetworkClientId`, `NetworkConfiguration['nativeTokenName']`)
- Transient state (`NetworkMetadata['status']`)
- Inherent/immutable properties that affect behavior (e.g. `NetworkMetadata['EIPS']`)
- Configurable/mutable properties that affect behavior (e.g. `NetworkConfiguration['rpcEndpoints]`)

mcmire replied:
Good idea to rename `NetworkMetadata`. I have never quite liked that name.

Also a good idea to think about how state properties ought to be categorized. Knowing that some state properties are transient and can easily be recreated, and that some properties are expected to be immutable is important.

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -350,44 +484,76 @@ export type NetworkControllerMessenger = RestrictedControllerMessenger<
+        nativeTokenName: NetworksTicker[infuraNetworkType],
+        rpcEndpoints: [
+          {
+            name: `Infura ${NetworkNickname[infuraNetworkType] as string}`,
+            networkClientId: infuraNetworkType,
+            type: RpcEndpointType.Infura,
+            url: rpcEndpointUrl,
+          },
+        ],
+      };
 
-type MetaMetricsEventPayload = {
-  event: string;
-  category: string;
-  referrer?: { url: string };
-  actionId?: number;
-  environmentType?: string;
-  properties?: unknown;
-  sensitiveProperties?: unknown;
-  revenue?: number;
-  currency?: string;
-  value?: number;
-};
+      return { ...obj, [chainId]: networkConfiguration };
+    },
+    {},
+  ) as Record<Hex, NetworkConfiguration>;
+}
```
It looks like we can get rid of the type asssertions here.

```suggestion
  return Object.values(InfuraNetworkType).reduce<Record<Hex, NetworkConfiguration>>(
    (obj, infuraNetworkType) => {
      const chainId = ChainId[infuraNetworkType];
      const rpcEndpointUrl = `https://${infuraNetworkType}.infura.io/v3/{infuraProjectId}`;

      const networkConfiguration: NetworkConfiguration = {
        chainId,
        defaultRpcEndpointUrl: rpcEndpointUrl,
        name: NetworkNickname[infuraNetworkType],
        nativeTokenName: NetworksTicker[infuraNetworkType],
        rpcEndpoints: [
          {
            name: `Infura ${NetworkNickname[infuraNetworkType]}`,
            networkClientId: infuraNetworkType,
            type: RpcEndpointType.Infura,
            url: rpcEndpointUrl,
          },
        ],
      };

      return { ...obj, [chainId]: networkConfiguration };
    },
    {},
  );
}
```

If you want to be strict about `obj`'s type during iteration, reverting line 502 from the above to the following also works:
```ts
    (obj: Partial<Record<Hex, NetworkConfiguration>>, infuraNetworkType) => {
```

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+          operation.value.networkClientId,
+        );
+        networkClient.destroy();
+        delete autoManagedNetworkClientRegistry[
+          networkClient.configuration.type
+        ][operation.value.networkClientId];
```
This and other `delete` operations here will probably trigger the ["prototype-polluting assignment"](https://github.com/MetaMask/core/pull/4041/files#diff-78fe44eeec5755b58fbeef2367e9a09861f63ba1d71dacd6494060d609187e76L110-L129) CodeQL alert. `isSafeDynamicKey`

```suggestion
        if (![networkClient.configuration.type, operation.value.networkClientId].every((e) => isSafeDynamicKey(e))) {
          continue;
        }
        delete autoManagedNetworkClientRegistry[
          networkClient.configuration.type
        ][operation.value.networkClientId];
```

mcmire replied:
I believe this only applies if the key you're using comes from the outside, right? But in this case I'm controlling what `networkClient.configuration.type` and `operation.value.networkClientId` can be. So perhaps this isn't needed?

MajorLift replied:
I think the static code analysis just flags any dynamic string being used for property access, which should apply here. But I don't think there's any actual risk of a pollution/injection attack here. We can always come back and add `isSafeDynamicKey` if this is flagged.

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -396,25 +562,149 @@ type AutoManagedCustomNetworkClientRegistry = Record<
 
+/**
+ * Determines whether the given URL is valid by attempting to parse it.
+ *
```
Could this be useful for other controllers?

```suggestion
 *
 * TODO: Move to @metamask/utils.
 *
```

mcmire replied:
Maybe. I haven't seen this anywhere else, and this also relies on a package to do the parsing (`uri-js`), which we may not want to introduce as a dependency to `@metamask/utils`.

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1253,108 +1980,77 @@ export class NetworkController extends BaseController<
+    const sortedChainIds = Object.keys(
+      this.state.networkConfigurationsByChainId,
+    )
+      .map(hexToBigInt)
+      .sort()
+      .map(toHex);
```
nit: These are currently safe because `hexToBigInt`, `toHex` are unary, but might break if we add additional option arguments to these functions in the future. 

```suggestion
      .map((e) => hexToBigInt(e))
      .sort()
      .map((e) => toHex(e));
```

mcmire replied:
Good shout. I'm okay with this — it seems unlikely that we would add new arguments any time soon, but this is an easy fix if that happens.

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -63,32 +73,178 @@ export type NetworkMetadata = {
- * Custom RPC network information
+ * The type of an RPC endpoint.
+ *
+ * @see CustomRpcEndpoint
+ * @see InfuraRpcEndpoint
+ */
```
nit:
```suggestion
 * @see {@link CustomRpcEndpoint}
 * @see {@link InfuraRpcEndpoint}
 */
```

mcmire replied:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -63,32 +73,178 @@ export type NetworkMetadata = {
+ * An RPC endpoint is a reference to a server which fronts an EVM chain. There
+ * are two varieties of RPC endpoints: Infura and custom.
+ *
+ * @see CustomRpcEndpoint
+ * @see InfuraRpcEndpoint
+ */
```
nit:
```suggestion
 * @see {@link CustomRpcEndpoint}
 * @see {@link InfuraRpcEndpoint}
 */
```

mcmire replied:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -63,32 +73,178 @@ export type NetworkMetadata = {
+ * Custom RPC endpoints do not need a `networkClientId` property because it is
+ * assumed that they have not already been added and are not represented by
+ * network clients yet.
+ *
```
nit:
```suggestion
```

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -186,26 +295,35 @@ export type CustomNetworkClientId = string;
+   * The registry of networks and corresponding RPC endpoints that the
+   * controller can use to make requests for various chains.
+   *
+   * @see NetworkConfiguration
```
```suggestion
   * @see {@link NetworkConfiguration}
```

mcmire replied:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+   * the RPC endpoints which front that chain.
+   * @returns The newly added network configuration.
+   * @throws if any part of `fields` would produce invalid state.
+   * @see NetworkConfiguration
```
nit:
```suggestion
   * @see {@link NetworkConfiguration}
```

mcmire replied:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+      );
+      const infuraNetworkNickname = NetworkNickname[infuraNetworkName];
+      const infuraChainId = ChainId[infuraNetworkName];
+      if (chainId !== infuraChainId) {
+        throw new Error(
+          // This is a string.
+          // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+          `Cannot add network with chain ID ${chainId} and Infura RPC endpoint for '${infuraNetworkNickname}' which represents ${infuraChainId}, as the two conflict`,
```
Looks like this might have been a false alarm. I'm not seeing any error messages here. Every so often I need to restart the typescript server in vscode to get the correct output.

```suggestion
          `Cannot add network with chain ID ${chainId} and Infura RPC endpoint for '${infuraNetworkNickname}' which represents ${infuraChainId}, as the two conflict`,
```

mcmire replied:
Hmm, interesting. I've noticed that having old TypeScript build files around can confuse the linter. I'll try clearing those out and re-running `yarn lint`.

MajorLift replied:
No need, I'm seeing the errors now! Looks like there was something about by local config that was off. Sorry for the confusion. 

We probably should investigate why `restrict-template-expressions` raises so many false positives for `any` though.

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+   * @returns The updated network configuration.
+   * @throws if `chainId` does not refer to an existing network configuration,
+   * or if any part of `fields` would produce invalid state.
+   * @see NetworkConfiguration
```
nit:
```suggestion
   * @see {@link NetworkConfiguration}
```

mcmire replied:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+        )
+      ) {
+        throw new Error(
+          `Cannot update network: RPC endpoint '${
+            // This is a string.
+            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+            url
+          }' refers to network client ${inspect(
```
Also not seeing an error here.

```suggestion
          `Cannot update network: RPC endpoint '${url}' refers to network client ${inspect(
```

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+      const infuraNetworkNickname = NetworkNickname[infuraNetworkName];
+      const infuraChainId = ChainId[infuraNetworkName];
+      if (newChainId !== infuraChainId) {
+        throw new Error(
+          // This is a string.
+          // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+          `Cannot update network with chain ID ${newChainId} and Infura RPC endpoint for '${infuraNetworkNickname}' which represents ${infuraChainId}, as the two conflict`,
+        );
```
Not seeing an error here.

```suggestion
        throw new Error(
          `Cannot update network with chain ID ${newChainId} and Infura RPC endpoint for '${infuraNetworkNickname}' which represents ${infuraChainId}, as the two conflict`,
        );
```

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+   * @param chainId - The chain ID associated with an existing network.
+   * @throws if `chainId` does not refer to an existing network configuration,
+   * or if the currently selected network is being removed.
+   * @see NetworkConfiguration
```
nit:
```suggestion
   * @see {@link NetworkConfiguration}
```

mcmire replied:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).

MajorLift commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -991,201 +1294,627 @@ export class NetworkController extends BaseController<
       this.#ensureAutoManagedNetworkClientRegistryPopulated();
-    const networkClientId = networkConfigurationId;
+
+    if (!isStrictHexString(newChainId) || !isSafeChainId(newChainId)) {
```
There are enough error messages here that I think it makes sense to focus on that in a separate ticket.

mcmire replied:
I ended up refactoring this anyway, because it was getting pretty difficult to manage and keep track of.

MajorLift reviewed (commented):
So far these comments are focused on the `NetworkController.ts` file. I'll continue looking at the rest.

mcmire commented:
@metamaskbot publish-preview

mcmire commented:
@metamaskbot publish-preview

MajorLift commented on tests/helpers.ts:
```diff
@@ -39,3 +41,48 @@ export async function advanceTime({
+      ...getKnownPropertyNames<keyof Type>(overrides),
+    ]),
+  ];
+  const object = keys.reduce((workingObject: Partial<Type>, key) => {
+    if (key in overrides) {
+      return { ...workingObject, [key]: overrides[key] };
+    } else if (key in defaults) {
+      return { ...workingObject, [key]: defaults[key]() };
+    }
+    return workingObject;
+  }, {}) as unknown as Type;
```
We can type this `reduce` with a generic parameter instead of using `as unknown as` .

```suggestion
  const object = keys.reduce<Type>((workingObject, key) => {
    if (key in overrides) {
      return { ...workingObject, [key]: overrides[key] };
    } else if (key in defaults) {
      return { ...workingObject, [key]: defaults[key]() };
    }
    return workingObject;
  }, {} as never);
```

mcmire replied:
Hmm, I like typing the "working object" as `Partial<Type>` and then typecasting the final result as `Type` instead of passing `Type` as a type parameter because this makes sense conceptually to me (i.e. it's not complete until it's complete). Now that I say that, I guess I don't need the `as unknown`:

``` suggestion
  const object = keys.reduce((workingObject: Partial<Type>, key) => {
    if (key in overrides) {
      return { ...workingObject, [key]: overrides[key] };
    } else if (key in defaults) {
      return { ...workingObject, [key]: defaults[key]() };
    }
    return workingObject;
  }, {}) as Type;
```

In any case, does this go against any of the guidelines you make in `contributor-docs`?

MajorLift replied:
I guess this depends on what we're trying to achieve by applying the `Partial` type, and whether there's a useful difference introduced by using `Partial<Type>` vs. `Type`.

1. `Partial` does not remove any keys from the index signature of `Type`. It only makes the corresponding values nullable (i.e. `| undefined`). 

`Partial<Type>` does not prevent unassigned keys from existing on `workingObject`. Applying `Partial` does not prevent any `key` value from being used for property access on `workingObject` at any point during the `reduce` loop. 

2. `Partial` only enforces that all keys may return `undefined`. It doesn't enforce that some specific set of keys must always return `undefined`. 

If we want to ensure that unassigned keys always return `undefined`, we'd need to write a more accurate but complex type, while keeping track of the iterated and remaining keys each loop e.g.

```ts
type WorkingObject = Omit<Type, IteratedKeys> & Partial<Pick<Type, RemainingKeys>>
```

3. `Partial` expands the probability space rather than helping us narrow it down.

Typing `workingObject` with `Type` restricts its index signature to the correct set of keys, and the assignable property values to the correct set of types. 

Replacing that with `Partial<Type>` expands the range of possibilities to include scenarios where different sets of keys are optional/required, and values are widened to be nullable in addition to whatever type they were previously restricted to.

* * *

I think that the constraints enforced by the `reduce` type parameter is good enough for most purposes, especially if it lets us avoid type assertion. The `@typescript-eslint/prefer-reduce-type-parameter` eslint rule operates on that principle, and it's on our list of rules to re-enable. I believe it could replace the `contributor-docs` entry we have about this topic.

That said, I'd definitely be interested in coming up with a better typing for `reduce` that distinguishes between in-progress and complete states for the working object.

mcmire replied:
Gotcha, thanks for the explanation. It seems that there are two ways to view this:

- You could argue that the working object is an incomplete version of the final version, so `Partial` communicates that it's possible that any of the keys in that object haven't been set yet as it's being built.
- You could also argue that widening the type inside of the `reduce` function is rarely valuable and that it is more valuable to assume that the object is complete all the time.

Despite what I just said above, I guess I don't have a real preference for this. I didn't realize that there was an ESLint rule for preferring the type parameter, so if that's true (and we can use it to remove some words from the contributor docs) then it makes me lean toward your suggestion.

MajorLift commented on packages/controller-utils/CHANGELOG.md:
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Added
+
+- Add `BlockExplorerUrl` for looking up the block explorer URL for any Infura network ([#4268](https://github.com/MetaMask/core/pull/4286))
+- Add `NetworkNickname` for looking up the common nickname for any Infura network ([#4268](https://github.com/MetaMask/core/pull/4286))
+- Add `Partialize` for making select keys in an object type optional ([#4268](https://github.com/MetaMask/core/pull/4286))
```
nit:
```suggestion
- Add `BlockExplorerUrl` object and type for looking up the block explorer URL of any Infura network ([#4268](https://github.com/MetaMask/core/pull/4286))
- Add `NetworkNickname` object and type for looking up the common nickname for any Infura network ([#4268](https://github.com/MetaMask/core/pull/4286))
- Add `Partialize` type for making select keys in an object type optional ([#4268](https://github.com/MetaMask/core/pull/4286))
```

bergeron commented on packages/network-controller/src/index.ts:
```diff
@@ -1,4 +1,38 @@
+export {
+  getDefaultNetworkControllerState,
+  knownKeysOf,
+  NetworkController,
```
Could `RpcEndpointType` be exported too?  I have some places it would be useful

```suggestion
  NetworkController,
  RpcEndpointType
```

mcmire replied:
Sure! Done.

mcmire commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -354,44 +488,77 @@ export type NetworkControllerMessenger = RestrictedControllerMessenger<
+ *
+ * @returns The default value for `networkConfigurationsByChainId`.
+ */
+function getDefaultNetworkConfigurationsByChainId(): Record<
```
Good suggestion. I'll have to do this in a different PR, since it will require changes across multiple packages.

mcmire commented:
New changes:

- `defaultRpcEndpointUrl` is now `defaultRpcEndpointIndex`
- `blockExplorerUrl` is now `blockExplorerUrls`; added `defaultBlockExplorerUrlIndex`
- The `name` property on RpcEndpoint is now optional

mcmire commented:
@metamaskbot publish-previews

mcmire commented:
@metamaskbot publish-preview

mcmire commented:
Items to do before merging:

- Resolve conflicts.
- The assets team requested that we add an option to `updateNetwork` which allows for removing the default RPC endpoint by making it possible to simultaneously switching the network. Currently we throw an error when the default RPC endpoint is being removed, but instead of throwing an error we can require a `newNetworkClientId` option to be passed instead.

mcmire commented:
I thought of something else that I should probably do now vs. later: if the URL of an RPC endpoint changes, but the `networkClientId` stays the same, then I'm not sure whether the controller will create a new network client or not. If it doesn't, then perhaps we should account for this (because then the URL associated with the RPC endpoint will be different from the actual URL used by the network client).

mcmire commented:
@metamaskbot publish-preview

mcmire commented:
Changes since last preview build:

- The `updateNetwork` method now throws an error when the RPC endpoint that is represented by the globally selected network is removed (otherwise, `selectedNetworkClientId` would have no corresponding RPC endpoint).
  - To prevent this error, you'll want to detect when an RPC endpoint is being removed and designate which RPC endpoint should be used for the globally selected network by passing a `replacementSelectedRpcEndpointIndex` option. This must point to an object in `rpcEndpoints`. `updateNetwork` will then automatically issue a network switch, and it will update both `selectedNetworkClientId` and `networkConfigurationsByChainId` at the same time to avoid creating a moment where the state is invalid.
  - Note that if an RPC endpoint is changed in such a way that necessitates recreating the network client for that RPC endpoint, `updateNetwork` will automatically use the new network client without having to specify `replacementSelectedRpcEndpointIndex`.
  - If you'd like to always pass `replacementSelectedRpcEndpointIndex`, there isn't a good way to get this without looping through `networkConfigurationsByChainId`. It can be done, but let me know if you need a more convenient way.
- `updateNetwork` now throws if you try to add an invalid Infura RPC endpoint or change any part of an Infura RPC endpoint (since they must have specific data at all times).
  - The `url` must match a specific string.
- `addNetwork` and `updateNetwork` now throw if `defaultBlockExplorerUrlIndex` is invalid.
- `updateNetwork` will now gracefully handle the case in which a `networkClientId` for a previously added RPC endpoint is unset (it will just fill it back in without creating a new network client)
- `updateNetwork` is also smarter about keeping track of RPC endpoints, so for instance you can shuffle around RPC endpoints and it won't change state or create any new network clients.

mcmire commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1063,205 +1482,427 @@ export class NetworkController extends BaseController<
+    ) {
+      await this.setActiveNetwork(rpcEndpointToSelect.networkClientId, {
+        updateState: (state) => {
+          this.#updateNetworkConfigurations({
```
This was a fun one to figure out. I want to ensure that state updates get combined so that we don't have a brief moment where we've changed `selectedNetworkClientId` but `networkConfigurationsByChainId` is now out of date. I solved this by passing a function that takes state and changes the `networkConfigurationsByChainId` property, and then I call at the point where I set `selectedNetworkClientId`. But it'd be cool if we had a better way of doing this.

salimtb commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1063,205 +1482,427 @@ export class NetworkController extends BaseController<
     }
 
+    const existingChainId = chainId;
+    const { chainId: newChainId, rpcEndpoints: setOfNewRpcEndpointFields } =
```
does this mean that only `chainId` and `rpcEndpoints` are editable ? we're not extracting name and ticker from the `fields`

mcmire replied:
All fields are editable. We destructure `chainId` and `rpcEndpoints` here so they're easier to work with on their own below, but the whole `fields` object ends up getting passed and updated in state, e.g.:

``` typescript
      this.update((state) => {
        this.#updateNetworkConfigurations({
          state,
          mode: 'update',
          networkFields: fields,
          networkConfigurationToPersist: updatedNetworkConfiguration,
          existingNetworkConfiguration,
        });
      });
```

salimtb commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -66,32 +77,189 @@ export type NetworkMetadata = {
+ * assumed that they are new and are not represented by network clients yet.
+ */
+type UpdateNetworkFields = Omit<NetworkConfiguration, 'rpcEndpoints'> & {
+  rpcEndpoints: (InfuraRpcEndpoint | UpdateNetworkCustomRpcEndpointFields)[];
```
does this mean that only rpcEndpoints is updatable ?

mcmire replied:
No, all properties are updatable. This type mixes in the same properties that `NetworkConfiguration` has, but overrides `rpcEndpoints` so that `networkClientId` is an optional property for each RPC endpoint instead of being required.

bergeron commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1063,205 +1481,427 @@ export class NetworkController extends BaseController<
+      throw new Error(
+        // False negative - this is a string.
+        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+        `Could not update network: Cannot update RPC endpoints in such a way that the selected network '${this.state.selectedNetworkClientId}' would be removed without a replacement. Choose a different RPC endpoint as the selected network via the \`replacementSelectedRpcEndpointIndex\` option.`,
```
This error is throwing in cases I don't think it should.  Example:

MM is currently on Mainnet, with a `selectedNetworkClientId` pointing to one of that chain's RPC endpoints.  I edit another chain, like changing the name or endpoints of Base network.

This line throws.  But I'm editing a different chain, which shouldn't conflict.   Wondering if I've got a bug somewhere, or this needs to check whether `selectedNetworkClientId` is on the chain being updated.

mcmire replied:
Hmm, you're right. This doesn't seem right to me either, now that I look at it again. I'll take another look today and try to get this fixed quickly.

mcmire replied:
Alright, I've made some changes that should hopefully fix the issue you're seeing.

mcmire replied:
This should be fixed in the latest preview build, but let me know if not.

mcmire commented:
@metamaskbot publish-preview

mcmire commented:
Pushed some updates to fix tests and ensure that we are testing `updateNetwork` we do not try to update a network configuration that has an RPC endpoint that is represented by the selected network client, unless we explicitly want to. This should guarantee that we don't have the same issue that Brian reported.

That should be it for updates unless we receive another bug report.

bergeron commented on packages/network-controller/src/index.ts:
```diff
@@ -1,4 +1,39 @@
+  NetworksMetadata,
+  NetworkState,
+  BlockTrackerProxy,
+  ProviderProxy,
```
Could these 2 be exported as well?  Otherwise I end up re-creating them in the client.  Useful for representing a network before it's been added.

```suggestion
  ProviderProxy,
  AddNetworkFields,
  UpdateNetworkFields,
```

mcmire replied:
Yup! Done.

mcmire commented:
@metamask publish-preview

mcmire commented:
@metamaskbot publish-preview

mcmire commented:
Updates since previous preview release:

- Added `AddNetworkFields` and `UpdateNetworkFields` types.

mikesposito commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1063,205 +1482,427 @@ export class NetworkController extends BaseController<
+    const networkClientOperations: NetworkClientOperation[] = [];
+
+    for (const newRpcEndpointFields of setOfNewRpcEndpointFields) {
+      const existingRpcEndpointForNoop =
```
The logic seems sound but as you mentioned it's quite bulky - I agree that moving these `CRUD` operations on Networks would help a lot, especially because we could then move all config validators to a more specific class, making `NetworkController` simpler

mikesposito commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1063,205 +1475,441 @@ export class NetworkController extends BaseController<
+    const autoManagedNetworkClientRegistry =
+      this.#ensureAutoManagedNetworkClientRegistryPopulated();
+
+    this.#validateNetworkFields({
```
I noticed that we call this method in two places only, but we need to pass `autoManagedNetworkClientRegistry` and both `mode` and `errorMessagePrefix`.

I feel like this gives us the ability to extend its validation capability, but at the expense of a higher complexity for the caller: do you think that we would be able to derive `autoManagedNetworkClientRegistry` and `errorMessagePrefix` based on the `mode` alone?

mcmire replied:
It's certainly possible to derive `errorMessagePrefix` inside of `#validateNetworkFields` instead of asking the caller to pass it. I've pushed commit [f25fb8c](https://github.com/MetaMask/core/pull/4286/commits/f25fb8c5693173b8147dba83aeb8152697af3619) to address this.

For `autoManagedNetworkClientRegistry`, we can't derive that, as we need it in `updateNetwork` ahead of time since we also need to pass it to `#registerClientsAsNeeded` and `#unRegisterClientsAsNeeded`.

mikesposito replied:
I see, that makes sense! Thanks for addressing `errorMessagePrefix`

mikesposito commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1063,205 +1475,441 @@ export class NetworkController extends BaseController<
+        networkClientOperations,
+      });
+
+    /*
+    console.log(
+      'replacementSelectedRpcEndpointIndex',
+      replacementSelectedRpcEndpointIndex,
+      'updatedNetworkConfiguration',
+      updatedNetworkConfiguration,
+      'networkClientOperations',
+      networkClientOperations,
+      'this.state.selectedNetworkClientId',
+      this.state.selectedNetworkClientId,
+    )
+    */
```
```suggestion
```

mcmire replied:
Oops, thank you :) Removed.

mikesposito commented on packages/controller-utils/src/types.ts:
```diff
@@ -87,3 +87,34 @@ export enum NetworksTicker {
+ * @template Type - The Record that you want to operate on.
+ * @template Key - The union of keys you want to make optional.
+ */
+export type Partialize<Type, Key extends keyof Type> = Omit<Type, Key> &
```
Perhaps we could add a `// TODO:` or open an issue so that we'll remember to move it sometime in the future?

mcmire commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -569,12 +966,13 @@ export class NetworkController extends BaseController<
-   * their identifier in the network client registry. This collection represents
-   * not only built-in networks but also any custom networks that consumers have
-   * added.
+   * Internally, the Infura and custom network clients are categorized by type
+   * so that when accessing either kind of network client, TypeScript knows
+   * which type to assign to the network client. For some cases it's more useful
+   * to be able to access network clients by ID instead of by type and then ID,
+   * so this function makes that possible.
```
Would adding this line be redundant given that we already have it under the `@returns` keyword? Or, is that something that you might expect `typedoc` to use but you don't tend to use yourself when you are reading JSDocs?

mikesposito replied:
Ah, I see now that was moved to the `@returns` - I think that's fine then!

mikesposito reviewed (approved):
LGTM!

Gudahtt commented:
Marking this as "DO-NOT-MERGE" until we are prepared to update this on both mobile and extension (e.g. until we have a draft PR somewhat-close to working)

mcmire commented:
Fixed conflicts.

mikesposito reviewed (approved):
Fixed conflicts. The issue addressed by [this PR](https://github.com/MetaMask/core/pull/4566) is handled differently here, so merged changes related to that have been discarded

bergeron commented:
I suspect this fix is needed in `SelectedNetworkController`: https://github.com/MetaMask/core/commit/1100b8fe33d2edc705cb79ce6f4dcad4d85f5652

Otherwise this code does not trigger because it expects `networkConfigurations` but receives `networkConfigurationsByChainId`.

![image](https://github.com/user-attachments/assets/dfa70356-edcb-4a9a-95f9-4fc19257202f)

bergeron commented:
I suspect a fix is needed in `SelectedNetworkController`.

This code to remove deleted network client IDs from dapp domains no longer triggers, since it  expects `networkConfigurations` but receives `networkConfigurationsByChainId`.

![image](https://github.com/user-attachments/assets/dfa70356-edcb-4a9a-95f9-4fc19257202f)

But also `path[1]` now represents a chain id instead of a network client id.  So I'm not sure how we'd access all the network client IDs that were deleted along with the chain.  And I think we'd also want to trigger this when RPC endpoints are deleted within a chain.

It may make sense for `SelectedNetworkController` to migrate from `Record<Domain, NetworkClientId>` to  `Record<Domain, ChainId>`.  I was thinking we'd do this in the future, but it may be easier to do now.

bergeron commented on packages/network-controller/src/index.ts:
```diff
@@ -1,5 +1,42 @@
+  BlockTrackerProxy,
+  ProviderProxy,
+  AddNetworkFields,
+  UpdateNetworkFields,
```
```suggestion
  UpdateNetworkFields,
  RpcEndpoint,
```

`RpcEndpoint` would be convenient to export

bergeron commented on packages/network-controller/src/NetworkController.ts:
```diff
@@ -1353,6 +1953,458 @@ export class NetworkController extends BaseController<
+
+    this.#networkConfigurationsByNetworkClientId =
+      buildNetworkConfigurationsByNetworkClientId(
+        this.state.networkConfigurationsByChainId,
```
```suggestion
        state.networkConfigurationsByChainId,
```

I think this fixes a sneaky bug in the context of adding or updating a network.

The network is added to `state.networkConfigurationsByChainId` a few lines above.  But it's not yet added to `this.state.networkConfigurationsByChainId` at this point, because we're still inside an update function.  Without this fix, the new network does not get passed to `buildNetworkConfigurationsByNetworkClientId`.

![Screenshot 2024-08-26 at 8 35 38 PM](https://github.com/user-attachments/assets/6248a170-916b-407d-9cdd-0982e5386b6c)

bergeron replied:
I committed this change so I can publish a preview version containing it

bergeron replied:
Had to take another pass and fix it this way instead: https://github.com/MetaMask/core/pull/4286/commits/02e7ff99c640127d9b08c80f0951f1141c16b3c6

bergeron commented:
@metamaskbot publish-preview

bergeron commented:
@mikesposito The only remaining thing I can't figure out, is an error `Cannot perform 'get' on a proxy that has been revoked` thrown by `TokenDetectionController` when switching chains.  I don't see anything wrong with this pattern, but the `networkConfiguration` it receives here has `[[IsRevoked]]: true`

![Screenshot 2024-08-29 at 12 20 56 PM](https://github.com/user-attachments/assets/04e8f5d5-3bb1-4d0c-af60-857c1d986628)

bergeron commented:
@metamaskbot publish-preview

mikesposito commented:
> Edit: I think I've fixed it in https://github.com/MetaMask/core/commit/02e7ff99c640127d9b08c80f0951f1141c16b3c6 by performing buildNetworkConfigurationsByNetworkClientId after the call to this.update instead of inside it. This satisfies both criteria. That we're including the added/updated/removed network in the call to build...(), but it's also not revoked due to being a reference to draft state.

Thanks for taking care of that, the fix makes sense.

mikesposito commented:
@bergeron CI fixed here: https://github.com/MetaMask/core/pull/4638

mikesposito reviewed (approved):
Changes from @mcmire and @bergeron look great.

Since I made some changes as well, it would be nice to have an additional approval before merging this

mikesposito commented on packages/selected-network-controller/src/SelectedNetworkController.ts:
```diff
@@ -191,24 +192,21 @@ export class SelectedNetworkController extends BaseController<
-            );
-          }
-        });
+      (availableNetworkClientIds) => {
+        // if a network is updated or removed, update the networkClientId for all domains
+        // that were using it to the selected network client id
+        const { selectedNetworkClientId } = this.messagingSystem.call(
+          'NetworkController:getState',
+        );
+        Object.entries(this.state.domains).forEach(
+          ([domain, networkClientIdForDomain]) => {
+            if (!availableNetworkClientIds.includes(networkClientIdForDomain)) {
+              this.setNetworkClientIdForDomain(domain, selectedNetworkClientId);
+            }
+          },
+        );
       },
+      selectAvailableNetworkClientIds,
```
See: https://github.com/MetaMask/core/pull/4638

bergeron commented:
This is ready to go on extension.  Mobile is still in progress, but extension needs to go first
